### PR TITLE
DM-43060: Write CachingLimitedButler

### PR DESCRIPTION
**Pull request by mfisherlevine**
## Checklist

- [x] ran Jenkins
- [x] added a release note for user-visible changes to `doc/changes`

### CODE REVIEW

**Review by TallJimbo — APPROVED** [review]
Given that we pair-coded this, it shouldn't be a surprise that I'm happy with it.

The lack of test coverage is a problem.  I don't think it would be hard to add some here - especially if you add some debug logging as we've discussed to fire when the cache is used; you can then trap those warnings in the tests.  But I also know that it might be a big ask for you to do it _right now_ (as it would be for me to do them).  If you aren't up for adding some tests, could you at least create a ticket for that?

I would also appreciate a ticket for handling `getDeferred` vs. caching; I think that'd be best handled by refactoring `DeferredDatasetHandle` to have a reference back to its butler, so your override to `get` can be used, but I don't think your use case needs it done.

**Comment by mfisherlevine** [comment]
Of course, and thank you!

[Ticket for tests](https://rubinobs.atlassian.net/browse/DM-43353)
[Ticket for `getDeferred`](https://rubinobs.atlassian.net/browse/DM-43354)

**Comment by mfisherlevine** [comment]
I just added some debug logs to show the caching, ahead of the testing.

**Comment by mfisherlevine** [comment]
I just added some debug logs to show the caching, ahead of the testing.

**Comment by timj** [comment]
It looks like the caching butler is not even imported anywhere. Can we have a short test that imports it at least?

File: python/lsst/pipe/base/caching_limited_butler.py [hunk 1]
```diff
@@ -0,0 +1,194 @@
...
+        obj = self._wrapped.get(ref, parameters=parameters, storageClass=storageClass)
+        if ref.datasetType.name in self._cache_on_get and not parameters:
+            # and not parameters is to make sure we don't cache sub-images etc
+            self._cache[ref.datasetType.name] = (
+                ref.id,
+                InMemoryDatasetHandle(
+                    obj,
+                    storageClass=storageClass,
+                    dataId=ref.dataId,
+                    copy=ref.datasetType.name not in self._no_copy_on_cache,
```
Comment (andy-slac) [inline python/lsst/pipe/base/caching_limited_butler.py:121]: I understand that `copy` flag only matters on `handle.get`? So if you save this handle and return `obj`, the task could modify `obj` which will also modify contents of the `handle`? Probably need to copy `obj` before storing it in handle (unless `copy` is false)?

**Comment by mfisherlevine** [comment]
I have pushed an _extremely_ minimal test that the class instantiates here, and linked the tests ticket in a `TODO` there. Let me know if that doesn't look OK.

File: python/lsst/pipe/base/caching_limited_butler.py (hunk 1 above)
Comment (mfisherlevine) [reply python/lsst/pipe/base/caching_limited_butler.py:121]: I'm actually not quite sure. @TallJimbo could you comment, do you think?

Comment (mfisherlevine) [reply python/lsst/pipe/base/caching_limited_butler.py]: We resolved this OOB. Changes pushed and Jenkins running.

**Pull request merged**